crates/sui-core/src/authority/shared_object_version_manager.rs

```diff
@@ -190,7 +228,17 @@ fn assign_versions_for_certificate(
             }
         })
         .for_each(|(id, version)| {
-            shared_input_next_versions.insert(id, version);
+            assert!(
+                !txn_cancelled,
+                "Cancelled transactions should not update shared objects"
```

Comment:
i think it might be clearer to skip the entire block starting at line 219 if the tx is cancelled. iow this assertion is loop-invariant so why not hoist it

Reply:
Good point. Done. I think I might created `txn_cancelled` just for this purpose.